### PR TITLE
Fix: check definition of addon whether is conflict 

### PR DESCRIPTION
**Pull request by wangyikewxgm**
1. fix bug check an  addon whether been used when disable.
2. add check definitions of an addon whether exist.
3. fix check addon system requirement special case bug: `actual: v1.5.0-beta.1` require:  `>=1.5.0` res: `ture`

### Description of your changes

Fixes #

I have:

- [x] Read and followed KubeVela's [contribution process](https://github.com/kubevela/kubevela/blob/master/contribute/create-pull-request.md).
- [ ] [Related Docs](https://github.com/kubevela/kubevela.io) updated properly. In a new feature or configuration option, an update to the documentation is necessary. 
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### CODE REVIEW

File: references/cli/addon.go [hunk 1]
```diff
@@ -323,6 +326,7 @@ non-empty new arg
 	}
 	cmd.Flags().StringVarP(&addonVersion, "version", "v", "", "specify the addon version to upgrade")
 	cmd.Flags().BoolVarP(&skipValidate, "skip-version-validating", "s", false, "skip validating system version requirement")
+	cmd.Flags().BoolVarP(&overrideDefs, "override-defs", "o", false, "override already exist definitions within addon")
```
Comment (wonderflow) [inline references/cli/addon.go:329]:
```suggestion
	cmd.Flags().BoolVarP(&overrideDefs, "override", "o", false, "override existing definitions if conflict with those contained in this addon")
```

File: references/cli/addon.go [hunk 2]
```diff
@@ -203,6 +205,7 @@ func NewAddonEnableCommand(c common.Args, ioStream cmdutil.IOStreams) *cobra.Com
 	cmd.Flags().StringVarP(&addonVersion, "version", "v", "", "specify the addon version to enable")
 	cmd.Flags().StringVarP(&addonClusters, types.ClustersArg, "c", "", "specify the runtime-clusters to enable")
 	cmd.Flags().BoolVarP(&skipValidate, "skip-version-validating", "s", false, "skip validating system version requirement")
+	cmd.Flags().BoolVarP(&overrideDefs, "override-defs", "o", false, "override already exist definitions within addon")
```
Comment (wonderflow) [inline references/cli/addon.go:208]:
```suggestion
	cmd.Flags().BoolVarP(&overrideDefs, "override", "o", false, "override existing definitions if conflict with those contained in this addon")
```

Comment (wonderflow) [reply references/cli/addon.go:208]: if you're desigining other override options in the future, I would suggest the option name to be "override-definitions" and with no short args("o").

File: pkg/addon/addon.go [hunk 3]
```diff
@@ -1069,6 +1070,16 @@ func (h *Installer) dispatchAddonResource(addon *InstallPackage) error {
 		return errors.Wrap(err, "render addon definitions fail")
 	}
 
+	if !h.overrideDefs {
+		existDefs, err := checkConflictDefs(h.ctx, h.cli, defs, *app)
+		if err != nil {
+			return err
+		}
+		if len(existDefs) != 0 {
+			return fmt.Errorf(`definitions: %s in this addon already exist, if you want override them, Please use add "--override-defs" to re-enable or re-upgrade`, existDefs)
```
Comment (wonderflow) [inline pkg/addon/addon.go:1079]:
```suggestion
			return fmt.Errorf(`definitions: %s in this addon already exist, if you want override them, Please use arguments "--override" to enable`, existDefs)
```

I wonder if the case really exist when users are upgrading addons?

Comment (wonderflow) [reply pkg/addon/addon.go:1079]: Shall we update an addon with new definitions add?

File: pkg/addon/addon.go [hunk 4]
```diff
@@ -1286,26 +1297,30 @@ func checkSemVer(actual string, require string) (bool, error) {
...
 		return true, nil
 	}
 	if strings.Contains(actual, "-") && !strings.Contains(require, "-") {
-		smeVer := strings.TrimPrefix(actual[:strings.Index(actual, "-")], "v")
-		v, err := semver.NewVersion(smeVer)
+		semVer := strings.TrimPrefix(actual[:strings.Index(actual, "-")], "v")
+		if strings.Contains(require, ">=") && require[strings.Index(require, "=")+1:] == semVer {
+			// for case: `actual` is 1.5.0-beta.1 require is >=`1.5.0`
+			return false, nil
+		}
```
Comment (wonderflow) [inline pkg/addon/addon.go:1320]: are there any new test cases to cover it?

Comment (wonderflow) [reply pkg/addon/addon.go:1320]: sorry ,found it

File: pkg/addon/addon_test.go [hunk 5]
```diff
@@ -945,6 +945,11 @@ func TestCheckSemVer(t *testing.T) {
 			require: ">=v1.2.4-beta.3",
 			res:     false,
 		},
+		{
+			actual:  "1.5.0-beta.2",
+			require: ">=1.5.0",
+			res:     false,
+		},
```
Comment (wonderflow) [inline pkg/addon/addon_test.go:952]:
```suggestion
		},
		{
			actual:  "1.5.0-alpha.2",
			require: ">=1.5.0",
			res:     false,
		},		
```

File: pkg/addon/addon.go (hunk 3 above)
Comment (wangyikewxgm) [reply pkg/addon/addon.go:1079]: I think addon developer cannot avoid this situation in the future.

Comment (wonderflow) [reply pkg/addon/addon.go:1079]: can we notify what exsiting addon contain this definition? It's user friendly:

···
definition: %s in this addon already exist in %s ...
···

File: pkg/addon/utils.go [hunk 6]
```diff
@@ -409,3 +439,21 @@ func generateAnnotation(meta *Meta) map[string]string {
 func isErrorCueRenderPathNotFound(err error, path string) bool {
 	return err.Error() == fmt.Sprintf("var(path=%s) not exist", path)
 }
+
+func checkConflictDefs(ctx context.Context, k8sClient client.Client, defs []*unstructured.Unstructured, app v1beta1.Application) ([]string, error) {
```
Comment (wonderflow) [inline pkg/addon/utils.go:443]:
```suggestion
func checkConflictDefs(ctx context.Context, k8sClient client.Client, defs []*unstructured.Unstructured, appName string) ([]string, error) {
```

File: pkg/addon/utils.go [hunk 7]
```diff
@@ -409,3 +439,21 @@ func generateAnnotation(meta *Meta) map[string]string {
...
+	var res []string
+	for _, def := range defs {
+		err := k8sClient.Get(ctx, client.ObjectKeyFromObject(def), def)
+		if err == nil {
+			owner := metav1.GetControllerOf(def)
+			if owner == nil || owner.Kind != v1beta1.ApplicationKind || owner.Name != app.Name {
+				// if addon not belong to an addon or addon name is another one, we should put them in result
```
Comment (wonderflow) [inline pkg/addon/utils.go:450]: the condition judgement is really clever! But the explanation here is confusing.

```suggestion
			if owner == nil || owner.Kind != v1beta1.ApplicationKind || owner.Name != app.Name {
				// owner is nil means the definition don't belong to any addons, it must conflict with us.
				// owner.Kind != v1beta1.ApplicationKind or name not equal, it also means the owner is not this addon.
```

BTW, it is possible that Definition has more than 1 owners , what if  you're checking the wrong one?

Comment (wangyikewxgm) [reply pkg/addon/utils.go:450]: Have consider this situation, because addon-app always is the controller owner, which can only be one.

Comment (wonderflow) [reply pkg/addon/utils.go:450]: I think it's better to check if this application is one of the addon

File: pkg/addon/utils.go (hunk 6 above)
Comment (wangyikewxgm) [reply pkg/addon/utils.go:443]: Maybe we need use  other information（annotations or labels）about this app in the future？

File: pkg/addon/addon.go (hunk 3 above)
Comment (wangyikewxgm) [reply pkg/addon/addon.go:1079]: Yes, when conflict, will produce an error to tell what defs has exist.

File: pkg/addon/utils.go (hunk 6 above)
Comment (wonderflow) [reply pkg/addon/utils.go:443]: please  follow the KISS (keep it simple and stupid) principle

File: pkg/addon/utils.go (hunk 7 above)
Comment (wangyikewxgm) [reply pkg/addon/utils.go:450]: I think if the addon-app is controller owner of this def, that's mean the addon will modify at any time. Is it  fine to update it directly?

File: pkg/addon/utils.go [hunk 8]
```diff
@@ -409,3 +440,37 @@ func generateAnnotation(meta *Meta) map[string]string {
...
+
+func produceDefConflictError(conflictDefs map[string]string) error {
+	if len(conflictDefs) == 0 {
+		return nil
+	}
+	var errorInfo string
+	for _, s := range conflictDefs {
+		errorInfo += s
+	}
+	errorInfo += "if you want override them, Please use arguments '--override' to enable \n"
```
Comment (wonderflow) [inline pkg/addon/utils.go:474]:
```suggestion
	errorInfo += "if you want override them, please use argument '--override' to enable \n"
```

**Review by wonderflow — APPROVED** [review]
LGTM after nits

File: pkg/addon/addon.go (hunk 3 above)
Comment (wangyikewxgm) [reply pkg/addon/addon.go]: The print result is 
```
definition: helm already exist and not belong to any addon 
definition: kustomize in this addon already exist in fluxcd 
if you want override them, Please use arguments '--override' to enable

File: references/cli/addon.go [hunk 9]
```diff
@@ -203,6 +205,7 @@ func NewAddonEnableCommand(c common.Args, ioStream cmdutil.IOStreams) *cobra.Com
 	cmd.Flags().StringVarP(&addonVersion, "version", "v", "", "specify the addon version to enable")
 	cmd.Flags().StringVarP(&addonClusters, types.ClustersArg, "c", "", "specify the runtime-clusters to enable")
 	cmd.Flags().BoolVarP(&skipValidate, "skip-version-validating", "s", false, "skip validating system version requirement")
+	cmd.Flags().BoolVarP(&overrideDefs, "override-definitions", "", false, "override existing definitions if conflict with those contained in this addon")
 	return cmd
```
Comment (Somefive) [inline references/cli/addon.go:209]: What is this args for? Is there any cases for it?

**Review by Somefive — APPROVED** [review]

**Pull request merged**